### ISSUE TITLE
Consumer Message Lagging on High CPU Load

### ISSUE THREAD

**Issue by kl-tie** [dotnetcore/CAP#1587]
Hi, there was an issue with the message lag when our consumer app was currently having high CPU load on other processes that sometimes take up to 100% for a long period. The consumers stop consuming the message, and the message lag has been keep increasing until we have to restart the app. The consumers in our app are critical and need to be highly prioritized.

Could we make the consumers' thread have the highest priority than the other processes?

Our setups:
- ASP.NET Core 6.0
- DotNetCore.CAP 7.2.0
- Redpanda 24.1.9 as Kafka broker
- PostgreSQL 16

Reproduceable repo: https://github.com/kl-tie/dotnetcore-cap-test

Steps to reproduce:
1. Clone the repo.
2. Build the Docker image using the script inside `docker-build.bat`.
3. Run Docker Compose file using `docker-compose.yml`.
4. Open http://localhost:8080/topics/transaction?p=-1&s=50&o=-1#consumers for monitoring the message lag.
5. Open http://localhost:1234/swagger/index.html and hit `/api/stress-test/cpu` endpoint to simulate the high CPU usage.
6. The app will have 100% CPU utilization and all incoming messages will be stop processed. You can use the included `DotNetCap.Publisher` console app to produce a message.

**Comment by yang-xiaodong** [comment]
@kl-tie Please upgrade to version 8.1.0+ to see if the problem still exists?

**Comment by kl-tie** [comment]
@yang-xiaodong Hi, we've upgraded the repo with .NET 8 and CAP packages to 8.2.0, and the problem still persists on high CPU load (~100%): https://github.com/kl-tie/dotnetcore-cap-test/tree/cap-ver-8

Setups:
ASP.NET Core 8.0
DotNetCore.CAP 8.2.0
Redpanda 24.1.9 as Kafka broker
PostgreSQL 16

We also have edited the reproduceable steps above to include the PostgreSQL instance in Docker Compose so you don't need to setup an external PostgreSQL instance.

**Comment by kl-tie** [comment]
Hi @yang-xiaodong, thank you for testing the case. As you said, we are already aware that this is not the problem in the CAP itself, hence why we asked on whether it is possible to prioritize the CAP's thread processes.

At this point, we understand that the thread processes are scheduled by .NET's `TaskScheduler` and `ThreadPool` and it seems the priority cannot be explicitly set. However, when this happens, we couldn't really get a clear picture of the main issue of what causing the consumer processes stopped.

So now we would like to ask whether it is possible to throw a consistent exception or maybe warning message in the app's output, as you can see from the test results, only 1 out of 5 tests that produces a warning message regarding timeout polling interval. What we would like to have is a consistent message that we could use for our alert metrics, so we would immediately know when the CAP stop consuming from the application level.

**Comment by yang-xiaodong** [comment]
@kl-tie 
Based on your description, I think you are wondering if Grafana's observation of
CPU can see some warning message in the output console when there is a high load thus responding quickly.

The fact that CAP itself doesn't know if consumption has stopped is largely dependent on the interface provided by the client driver, and we don't check the current LAG to report such warnings.

I think it's also possible that there's some kind of [errors](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafkacpp_8h.html#a4c6b7af48c215724c323c60ea4080dbf) that stops the consumer and isn't included in our RetriableErrorCodes, which needs to be investigated further!

**Comment by kl-tie** [comment]
@yang-xiaodong Regarding the Grafana, we already setup our resources monitoring including CPU usage and alerting with Grafana. However, since our consumer apps might have high load during peak times. But then again, we would like to have alert on the consumer level itself, especially when there is an indication of consumers stop processing messages.

After taking a quick glance on the librdkafka docs you've mentioned above, I'm not sure we could get an error message regarding this case. However, what do you think if you put some kind of timer object like `Stopwatch` combined with event or `Observable` to monitor the consume process when it is invoking the `Confluent.Kafka` consume API?

**Comment by yang-xiaodong** [comment]
The problem is that monitor process may also stop

**Comment by kl-tie** [comment]
Yeah, you're right, the other processes in consumer apps will also have to wait for available threads.

**Comment by yang-xiaodong** [comment]
@APMIFEMA 

We have added more retryable error codes to the `RetriableErrorCodes` configuration, as follows:

```
ErrorCode.Local_Retry,
ErrorCode.RequestTimedOut,
ErrorCode.LeaderNotAvailable,
ErrorCode.NotLeaderForPartition,
ErrorCode.RebalanceInProgress,
ErrorCode.NotCoordinatorForGroup,
ErrorCode.NetworkException,
ErrorCode.GroupCoordinatorNotAvailable,
```

However, I believe that `Local_MaxPollExceeded` should not be added. This exception is caused by the `max.poll.interval.ms` configuration. This error is not suitable for retries because it indicates that the consumer’s behavior does not meet expectations. Typically, this means that the consumer logic needs to be optimized to ensure that the `poll()` method is called within the specified interval to prevent the consumer from being considered unresponsive.

Setting the `Local_MaxPollExceeded` error to be retried could cause the same message to be repeatedly polled by the consumer due to processing timeout.

In response to your error, I believe that the application design should be adjusted to ensure that each message is processed within Kafka’s maximum polling interval or that the `max.poll.interval.ms` configuration value is increased. 

As for timeouts caused by CPU overload, as discussed earlier, it seems there isn't a good solution for that scenario.

**Comment by yang-xiaodong** [comment]
Fixed in version 8.3.0
